### PR TITLE
Support End To End PagedAttention in JetStream

### PR DESCRIPTION
**Pull request by FanhaiLu1**
**Background:** 
PR #167 Added paged attention manager and kv cache manager. This PR supports end to end paged attention in JetStream. 

**Main Changes in this PR**

1. Added paged attention kernel in attention_kernel.py
2. Supported paged attention insert, decode in Engine.py
3. Refactor PageAttention manager 
4. Add unit test for kernel and manager

**Next Steps**

1. Tuning accuracy:  Current implement can output human readable output, but accuracy is low, only first few token are matches to dense attention
2. Improve performance: The kernel itself are almost same performance as dense attention. By applying bf16 calculation can boost the performance
4. Optimize the out jit compute
5. Elegant way to collect resource
6. Support quantization
7. Lazy cache update

**Current PR Output tokens example:** 

[304, 1284, 292, 367, 1476, 304, 1284, 714, 309, 310, 367, 29875, 18834, 29895, 29906, 29889, 29923, 310, 278, 367, 278, 1284, 4614, 17970, 2880, 310, 367, 579, 8024, 297, 25891, 29889, 306, 1311, 310, 1284, 12, 1284, 4634, 338, 304, 292, 6593, 310, 1196, 10379, 306, 29949, 6593, 1341, 263, 2834, 353, 29903, 12623, 29924, 278, 304, 350, 29889, 590, 1339, 29918, 2834, 6593, 310, 278, 310, 445, 338, 306, 508, 4658, 393, 338, 304, 437, 29973, 2023, 6593, 310, 306, 29915, 29879, 1914, 2834, 338, 263, 2462, 1383, 306, 4658, 592, 310, 10239, 306, 306, 273, 297, 306, 1774, 2834, 30010, 29871, 967, 6593, 310, 26093, 29880, 29889, 29914, 29968, 304, 4892, 29936, 29889, 338, 304, 306, 505, 2715, 29889, 372, 2191, 338, 278, 6593, 310, 29914, 2834, 306, 508, 29892, 322, 278, 29899, 306, 505, 29918, 278, 6593, 310, 306, 505, 6593, 29892, 306, 508, 367, 2834, 338, 304, 306, 723, 338, 16316, 310, 2834, 306, 505, 29889, 2794, 310, 306, 505, 304, 306, 306, 505, 263, 716, 29889, 29871, 29896, 29900, 29896, 29900, 30488, 29876, 29871, 29896, 29900, 306, 505, 263, 2462, 306, 505, 1063, 263, 29889, 29871, 29896, 29900, 29900, 526, 366, 508, 367, 29914, 29879, 2834, 338, 263, 29889, 29871, 29896, 29929, 29889, 29871, 29896, 29929, 29929, 29929, 29929, 29929, 29929, 29889, 306, 505, 263, 716, 15483, 322, 278, 1900, 310, 278, 6593, 310, 278, 6593, 310, 278, 6593, 310, 278, 1556, 310, 278, 1900, 29918, 278, 1900, 310, 278, 1556, 310, 278, 1900, 29899, 6707, 373, 278, 6593, 310, 12, 306, 626, 263, 306, 437, 29889, 29889, 306, 367, 592, 306, 626, 6593, 310, 304, 306, 4658, 306, 505, 263, 2462, 310, 278, 6593, 310, 278, 1900, 306, 505, 1063, 263, 716, 3088, 310, 278, 2446, 1629, 29899, 29900, 29889]

### CODE REVIEW

File: jetstream_pt/cache_manager.py [hunk 1]
```diff
@@ -663,18 +664,21 @@ def __init__(
...
     self.cache_k = cache_k
     self.cache_v = cache_v
+    self.page_attention_manager = page_attention_manager
     self.page_token_indices = page_token_indices
     self.sharding = sharding
     self.env = env
+    self.stacked = False
 
-  def update(self, key, value):
+  def update(self, key, value, layer_id=0):
```
Comment (qihqi) [inline jetstream_pt/cache_manager.py:681]: can pass layer_id from the constructor:
Each layer will get a different cache_manager instance anyways.

File: jetstream_pt/config.py [hunk 2]
```diff
@@ -139,6 +139,18 @@
     "size of top k used when sampling next token",
 )
 
+flags.DEFINE_integer(
```
Comment (qihqi) [inline jetstream_pt/config.py:142]: what is the boolean flag to enable / disable pagedattention?

File: jetstream_pt/config.py [hunk 3]
```diff
@@ -139,6 +139,18 @@
...
+    "total_num_pages",
```
Comment (qihqi) [inline jetstream_pt/config.py:143]: rename to pagedattention_total_num_pages

File: jetstream_pt/config.py [hunk 4]
```diff
@@ -139,6 +139,18 @@
...
+    0,
+    "total number of pages per layer for page attention",
+)
+
+flags.DEFINE_integer(
```
Comment (qihqi) [inline jetstream_pt/config.py:148]: pagedattention_page_size

File: jetstream_pt/environment.py [hunk 5]
```diff
@@ -140,6 +153,10 @@ def __init__(self, data: JetEngineEnvironmentData):
     self.testing = self._data.testing
     self.testing_seed = self._data.testing_seed
     self.ring_buffer = self._data.ring_buffer
+    self.page_attention = self._data.total_num_pages > 0
+    self.prefill_attention_kv_axis_names = (
```
Comment (qihqi) [inline jetstream_pt/environment.py:157]: not neccessary: there is a `__getattr__` method on this class that will get from `_data`

File: jetstream_pt/cache_manager.py (hunk 1 above)
Comment (FanhaiLu1) [reply jetstream_pt/cache_manager.py:681]: Good pint! Below are code to call the update in layer.py
`
 orig_keys, orig_values = cache.update(xk, xv, self.layer_id)
`


For page attention, the update didn't use layer_id (it's just a placer holder). I will refactor it later after this PR to isolate the potential for lazy update and normal update.

File: jetstream_pt/config.py (hunk 2 above)
Comment (FanhaiLu1) [reply jetstream_pt/config.py:142]: `self.page_attention` variable in JetEngineEnvironment are the flag. It's calculated by: 
`
  self.page_attention = self._data.total_num_pages > 0
`

File: jetstream_pt/environment.py (hunk 5 above)
Comment (FanhaiLu1) [reply jetstream_pt/environment.py:157]: Thanks, changed. Will change other similar behavior in later PR

Comment (FanhaiLu1) [reply jetstream_pt/environment.py:157]: In general, I feel __getattr__ is more prone:

`self.page_attention = self._data.paged_attention_total_num_pages > 0`

**vs**

`
    self.page_attention = (
        self.__getattr__("paged_attention_total_num_pages") > 0
    )
`

Comment (FanhaiLu1) [reply jetstream_pt/environment.py:157]: I would like to still keep attribute access other than self.__getattr__.  __getattr__ code more longer,  more prone and can not point the the variable definition.

File: jetstream_pt/config.py (hunk 4 above)
Comment (FanhaiLu1) [reply jetstream_pt/config.py:148]: Thanks, changed.

File: jetstream_pt/config.py (hunk 3 above)
Comment (FanhaiLu1) [reply jetstream_pt/config.py]: Changed.

File: jetstream_pt/environment.py (hunk 5 above)
Comment (qihqi) [reply jetstream_pt/environment.py:157]: if you do `self.paged_attention_total_num_pages` it will call `__getattr__` for you.

**Review by qihqi — APPROVED** [review]

File: jetstream_pt/environment.py (hunk 5 above)
Comment (FanhaiLu1) [reply jetstream_pt/environment.py:157]: Great, it works.

**Pull request merged**